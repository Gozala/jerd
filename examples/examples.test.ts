// Ok here we test things with fixtures I think

import { parse } from '../src/grammar';
import { typeDefine, typeEffect } from '../src/env';
import { presetEnv } from '../src/preset';
import fs from 'fs';
import { printToString } from '../src/printer';
import { declarationToPretty } from '../src/printTsLike';
import { declarationToString } from '../src/typeScriptPrinter';
import * as serializerRaw from 'jest-snapshot-serializer-raw';
import { SnapshotState } from 'jest-snapshot';

const toTest = [
    // yes
    'basic-effects.jd',
    'inference.jd',
    'generics.jd',
];

describe('Example files', () => {
    // beforeAll(() => {
    //     base = null;
    //     foundImages = [];
    //     // This makes the snapshot be put in the file corresponding to our fixture file,
    //     // not this snapshots.test.js file
    //     const updateSnapshot = (expect: any).getState().snapshotState
    //         ._updateSnapshot;
    //     const snapshotState = new SnapshotState(snapshotPath, {
    //         updateSnapshot,
    //     });
    //     (expect: any).setState({snapshotState});
    // });

    toTest.forEach((name) => {
        it(name, () => {
            // const snapshotPath = `__snapshots__/${name}.snap`;
            // const updateSnapshot = (expect as any).getState().snapshotState
            //     ._updateSnapshot;
            // // const currentState = (expect as any).getState().snapshotState;
            // // currentState._snapshotPath = snapshotPath;
            // const snapshotState = new SnapshotState(snapshotPath, {
            //     updateSnapshot,
            // } as any);
            // (expect as any).setState({ snapshotState });
            // // expect.addSnapshotSerializer(serializerRaw);

            const env = presetEnv();
            const raw = fs.readFileSync(__dirname + '/' + name, 'utf8');
            const parsed = parse(raw);
            parsed.forEach((item, i) => {
                if (item.type === 'effect') {
                    typeEffect(env, item);
                }
                if (item.type === 'define') {
                    const { term, hash } = typeDefine(env, item);
                    const jd = printToString(
                        declarationToPretty({ hash, size: 1, pos: 0 }, term),
                        100,
                    );
                    const ts = declarationToString(env, hash, term);
                    expect(serializerRaw.wrap(jd)).toMatchSnapshot(`${i}-jd`);
                    expect(serializerRaw.wrap(ts)).toMatchSnapshot(`${i}-ts`);
                }
            });

            // Saving the snapshot we just took, if need be
            // (expect as any).getState().snapshotState.save();
        });
    });
});
