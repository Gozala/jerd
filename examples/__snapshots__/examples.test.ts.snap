// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example files basic-effects.jd 1`] = `
// hello.jd
const hash_4062b9f2_0: () ={}> string = () => {
    "hello world";
}

// .ts
const hash_4062b9f2: () => string = () => {
  return "hello world";
};

// sayHi.jd
const hash_027c1aca_0: () ={}> void = () => {
    log(hash_4062b9f2_0());
}

// .ts
const hash_027c1aca: () => void = () => {
  return log(hash_4062b9f2());
};

// respondWith.jd
const hash_11e7ad08_0: (var_0_0) ={}> (() ={hash_a690a6f2_0}> T_0) ={}> T_0 = (
    responseValue_0: var_0_0,
) => (fn_1: () ={hash_a690a6f2_0}> T_0) => {
    handle! fn_1 {
        0(() => k_2) => <self>(responseValue_0 ++ ".")(() => k_2(responseValue_0)),
        1((v_2) => k_3) => {
            log(v_2);
            <self>(responseValue_0 ++ "-")(k_3);
        },
        pure(a_2) => a_2,
    };
}

// .ts
const hash_11e7ad08: (arg_0: T_0) => <T_0>(arg_0: (handlers: any, done: (result: T_0) => void) => void) => T_0 = responseValue_0 => fn_1 => {
  let handle_ret_0;
  handleSimpleShallow2("a690a6f2", fn_1, [(handlers, _, k_2) => {
    handle_ret_0 = hash_11e7ad08(responseValue_0 + ".")((handlers: any, done: (arg_0: T_0) => void) => {
      k_2(responseValue_0, handlers, done);
    });
  }, (handlers, v_2, k_3) => {
    handle_ret_0 = (() => {
      log(v_2);
      return hash_11e7ad08(responseValue_0 + "-")(k_3);
    })();
  }], (handlers, a_2) => {
    handle_ret_0 = a_2;
  });
  return handle_ret_0;
};

// farther.jd
const hash_22dc9717_0: (string) ={hash_a690a6f2_0}> string = (name_0: string) => {
    log("yes please " ++ name_0);
    raise!(hash_a690a6f2_0());
}

// .ts
const hash_22dc9717: (arg_0: string, handlers: any, done: (result: string) => void) => void = (name_0: string, handlers: any, done: (arg_0: string) => void) => {
  ((handlers, _ignored) => {
    raise(handlers, "a690a6f2", 0, null, (handlers, value) => done(handlers, value));
  })(handlers, log("yes please " + name_0));
};

// inner.jd
const hash_7f4162b0_0: (string) ={hash_a690a6f2_0}> void = (name_0: string) => {
    log(hash_22dc9717_0("Folks") ++ " from farther");
    log("getting");
    log(name_0);
    log(raise!(hash_a690a6f2_0()) ++ " and " ++ raise!(hash_a690a6f2_0()));
    log("And then " ++ raise!(hash_a690a6f2_0()));
    log("Dones");
}

// .ts
const hash_7f4162b0: (arg_0: string, handlers: any, done: (result: void) => void) => void = (name_0: string, handlers: any, done: (arg_0: void) => void) => {
  hash_22dc9717("Folks", handlers, (handlers, arg_0_6) => {
    ((handlers, arg_0_5) => {
      ((handlers, _ignored) => {
        ((handlers, _ignored) => {
          ((handlers, _ignored) => {
            raise(handlers, "a690a6f2", 0, null, (handlers, value) => ((handlers, arg_0_4) => {
              ((handlers, arg_0_3) => {
                raise(handlers, "a690a6f2", 0, null, (handlers, value) => ((handlers, arg_1_3) => {
                  ((handlers, arg_0_2) => {
                    ((handlers, _ignored) => {
                      raise(handlers, "a690a6f2", 0, null, (handlers, value) => ((handlers, arg_1_1) => {
                        ((handlers, arg_0_0) => {
                          ((handlers, _ignored) => {
                            done(handlers, log("Dones"));
                          })(handlers, log(arg_0_0));
                        })(handlers, "And then " + arg_1_1);
                      })(handlers, value));
                    })(handlers, log(arg_0_2));
                  })(handlers, arg_0_3 + arg_1_3);
                })(handlers, value));
              })(handlers, arg_0_4 + " and ");
            })(handlers, value));
          })(handlers, log(name_0));
        })(handlers, log("getting"));
      })(handlers, log(arg_0_5));
    })(handlers, arg_0_6 + " from farther");
  });
};

// test1.jd
const hash_297bf514_0: () ={}> T_0 = () => {
    hash_11e7ad08_0("<read>")(() => hash_7f4162b0_0("Yes"));
}

// .ts
const hash_297bf514: () => T_0 = () => {
  return hash_11e7ad08("<read>")((handlers: any, done: (arg_0: void) => void) => {
    hash_7f4162b0("Yes", handlers, done);
  });
};

// maybeRaise.jd
const hash_38749180_0: (bool) ={hash_a690a6f2_0}> string = (shouldRaise_0: bool) => {
    if shouldRaise_0 {
        raise!(hash_a690a6f2_0());
    } else {
        "didn't raise";
    };
}

// .ts
const hash_38749180: (arg_0: bool, handlers: any, done: (result: string) => void) => void = (shouldRaise_0: bool, handlers: any, done: (arg_0: string) => void) => {
  if (shouldRaise_0) {
    raise(handlers, "a690a6f2", 0, null, (handlers, value) => done(handlers, value));
  } else {
    done(handlers, "didn't raise");
  }
};

// test2.jd
const hash_35140570_0: () ={}> void = () => {
    hash_11e7ad08_0("<read>")(() => log(hash_38749180_0(true) ++ hash_38749180_0(false)));
}

// .ts
const hash_35140570: () => void = () => {
  return hash_11e7ad08("<read>")((handlers: any, done: (arg_0: void) => void) => {
    hash_38749180(true, handlers, (handlers, arg_0_1) => {
      hash_38749180(false, handlers, (handlers, arg_1_1) => {
        ((handlers, arg_0_0) => {
          done(handlers, log(arg_0_0));
        })(handlers, arg_0_1 + arg_1_1);
      });
    });
  });
};

// raiseIf.jd
const hash_65efa22a_0: (string) ={hash_a690a6f2_0}> string = (check_0: string) => {
    if raise!(hash_a690a6f2_0()) == check_0 {
        raise!(hash_a690a6f2_0());
    } else {
        "didn't raise";
    };
}

// .ts
const hash_65efa22a: (arg_0: string, handlers: any, done: (result: string) => void) => void = (check_0: string, handlers: any, done: (arg_0: string) => void) => {
  raise(handlers, "a690a6f2", 0, null, (handlers, value) => ((handlers, arg_0_0) => {
    ((handlers, cond) => {
      if (cond) {
        raise(handlers, "a690a6f2", 0, null, (handlers, value) => done(handlers, value));
      } else {
        done(handlers, "didn't raise");
      }
    })(handlers, arg_0_0 == check_0);
  })(handlers, value));
};

// test3.jd
const hash_7b912f5a_0: () ={}> void = () => {
    log(hash_11e7ad08_0("<read>")(() => hash_65efa22a_0("<read>")));
}

// .ts
const hash_7b912f5a: () => void = () => {
  return log(hash_11e7ad08("<read>")((handlers: any, done: (arg_0: string) => void) => {
    hash_65efa22a("<read>", handlers, done);
  }));
};

// letBind.jd
const hash_38ca1d5a_0: () ={hash_a690a6f2_0}> void = () => {
    const x_0 = raise!(hash_a690a6f2_0());
    log(x_0 ++ "Hello");
}

// .ts
const hash_38ca1d5a: (handlers: any, done: (result: void) => void) => void = (handlers: any, done: (arg_0: void) => void) => {
  raise(handlers, "a690a6f2", 0, null, (handlers, value) => ((handlers, x_0) => ((handlers, _ignored) => {
    done(handlers, log(x_0 + "Hello"));
  })(handlers))(handlers, value));
};

// test4.jd
const hash_69355794_0: () ={}> T_0 = () => {
    hash_11e7ad08_0("<read>")(hash_38ca1d5a_0);
}

// .ts
const hash_69355794: () => T_0 = () => {
  return hash_11e7ad08("<read>")(hash_38ca1d5a);
};

// .jd
hash_297bf514_0()

// .ts
hash_297bf514()

// .jd
hash_35140570_0()

// .ts
hash_35140570()

// .jd
hash_7b912f5a_0()

// .ts
hash_7b912f5a()

// .jd
hash_69355794_0()

// .ts
hash_69355794()
`;

exports[`Example files basics.jd 1`] = `
// .jd
true

// .ts
true

// .jd
"hi" == "hi"

// .ts
"hi" == "hi"

// freturn.jd
const hash_71396b2c_0: () ={}> string = () => "yes"

// .ts
const hash_71396b2c: () => string = () => "yes";

// .jd
hash_71396b2c_0() == "yes"

// .ts
hash_71396b2c() == "yes"

// arg.jd
const hash_66c69b52_0: (string) ={}> string = (arg_0: string) => arg_0 ++ "1"

// .ts
const hash_66c69b52: (arg_0: string) => string = arg_0 => arg_0 + "1";

// .jd
hash_66c69b52_0("2") == "21"

// .ts
hash_66c69b52("2") == "21"

// identity.jd
const hash_ef13ad00_0: (T_0) ={}> T_0 = (x_0: T_0) => x_0

// .ts
const hash_ef13ad00: <T_0>(arg_0: T_0) => T_0 = x_0 => x_0;

// .jd
hash_ef13ad00_0("5") ++ "4" == "54"

// .ts
hash_ef13ad00("5") + "4" == "54"

// provideString.jd
const hash_3971b373_0: (string) ={}> (() ={hash_762a49fa_0}> string) ={}> string = (
    responseValue_0: string,
) => (fn_1: () ={hash_762a49fa_0}> string) => {
    handle! fn_1 {
        0(() => k_2) => <self>(responseValue_0 ++ ".")(() => k_2(responseValue_0)),
        pure(a_2) => a_2,
    };
}

// .ts
const hash_3971b373: (arg_0: string) => (arg_0: (handlers: any, done: (result: string) => void) => void) => string = responseValue_0 => fn_1 => {
  let handle_ret_0;
  handleSimpleShallow2("762a49fa", fn_1, [(handlers, _, k_2) => {
    handle_ret_0 = hash_3971b373(responseValue_0 + ".")((handlers: any, done: (arg_0: string) => void) => {
      k_2(responseValue_0, handlers, done);
    });
  }], (handlers, a_2) => {
    handle_ret_0 = a_2;
  });
  return handle_ret_0;
};

// .jd
hash_3971b373_0("hi")(() => "m") == "m"

// .ts
hash_3971b373("hi")((handlers: any, done: (arg_0: string) => void) => {
  done(handlers, "m");
}) == "m"

// impure.jd
const hash_2a33607b_0: () ={hash_762a49fa_0}> string = () => "A"

// .ts
const hash_2a33607b: (handlers: any, done: (result: string) => void) => void = (handlers: any, done: (arg_0: string) => void) => {
  done(handlers, "A");
};

// .jd
hash_3971b373_0("hi")(hash_2a33607b_0) == "A"

// .ts
hash_3971b373("hi")(hash_2a33607b) == "A"

// pure.jd
const hash_4688a27a_0: () ={}> string = () => "AA"

// .ts
const hash_4688a27a: () => string = () => "AA";

// .jd
hash_3971b373_0("hi")(hash_4688a27a_0) == "AA"

// .ts
hash_3971b373("hi")((handlers: any, done: (arg_0: string) => void) => {
  done(handlers, hash_4688a27a());
}) == "AA"

// .jd
hash_3971b373_0("hi")(() => "m") == "m"

// .ts
hash_3971b373("hi")((handlers: any, done: (arg_0: string) => void) => {
  done(handlers, (() => "m")());
}) == "m"

// provideStringWithArg.jd
const hash_4dccbc5a_0: (string) ={}> (string, (string) ={hash_762a49fa_0}> string) ={}> string = (
    responseValue_0: string,
) => (passIn_1: string, fn_2: (string) ={hash_762a49fa_0}> string) => {
    handle! () => fn_2(passIn_1) {
        0(() => k_3) => hash_3971b373_0(responseValue_0 ++ ".")(() => k_3(responseValue_0)),
        pure(a_3) => a_3,
    };
}

// .ts
const hash_4dccbc5a: (arg_0: string) => (arg_0: string, arg_1: (arg_0: string, handlers: any, done: (result: string) => void) => void) => string = responseValue_0 => (passIn_1, fn_2) => {
  let handle_ret_0;
  handleSimpleShallow2("762a49fa", (handlers: any, done: (arg_0: string) => void) => {
    fn_2(passIn_1, handlers, done);
  }, [(handlers, _, k_3) => {
    handle_ret_0 = hash_3971b373(responseValue_0 + ".")((handlers: any, done: (arg_0: string) => void) => {
      k_3(responseValue_0, handlers, done);
    });
  }], (handlers, a_3) => {
    handle_ret_0 = a_3;
  });
  return handle_ret_0;
};

// .jd
hash_4dccbc5a_0("hi")("Passed in", (arg_0: string) => arg_0 ++ "-m") == "Passed in-m"

// .ts
hash_4dccbc5a("hi")("Passed in", (arg_0_0: string, handlers: any, done: (arg_0: string) => void) => {
  done(handlers, (arg_0 => arg_0 + "-m")(arg_0_0));
}) == "Passed in-m"
`;

exports[`Example files eff-paper.jd 1`] = `
// tee.jd
const hash_0b6f1408_0: (string) ={}> string = (x_0: string) => {
    log(x_0);
    x_0;
}

// .ts
const hash_0b6f1408: (arg_0: string) => string = x_0 => {
  log(x_0);
  return x_0;
};

// print.jd
const hash_596313f2_0: (string) ={hash_4e128638_0}> void = (x_0: string) => raise!(
    hash_4e128638_0(x_0),
)

// .ts
const hash_596313f2: (arg_0: string, handlers: any, done: (result: void) => void) => void = (x_0: string, handlers: any, done: (arg_0: void) => void) => {
  raise(handlers, "4e128638", 0, x_0, (handlers, value) => done(handlers, value));
};

// read.jd
const hash_79d200c4_0: () ={hash_762a49fa_0}> string = () => raise!(hash_762a49fa_0())

// .ts
const hash_79d200c4: (handlers: any, done: (result: string) => void) => void = (handlers: any, done: (arg_0: string) => void) => {
  raise(handlers, "762a49fa", 0, null, (handlers, value) => done(handlers, value));
};

// alwaysRead.jd
const hash_5d8d14a2_0: (string, () ={hash_762a49fa_0, hash_4e128638_0}> void) ={hash_4e128638_0}> void = (
    value_0: string,
    fn_1: () ={hash_762a49fa_0, hash_4e128638_0}> void,
) => {
    handle! fn_1 {0(() => k_2) => {
        <self>(value_0, () => k_2(value_0));
    }, pure(a_2) => a_2};
}

// .ts
const hash_5d8d14a2: (arg_0: string, arg_1: (handlers: any, done: (result: void) => void) => void, handlers: any, done: (result: void) => void) => void = (value_0: string, fn_1: (handlers: any, done: (result: void) => void) => void, handlers: any, done: (arg_0: void) => void) => {
  handleSimpleShallow2("762a49fa", fn_1, [(handlers, _, k_2) => {
    hash_5d8d14a2(value_0, (handlers: any, done: (arg_0: void) => void) => {
      k_2(value_0, handlers, done);
    }, handlers, done);
  }], (handlers, a_2) => {
    done(handlers, a_2);
  }, handlers);
};

// collect.jd
const hash_563322b8_0: (() ={hash_4e128638_0}> void) ={}> string = (
    fn_0: () ={hash_4e128638_0}> void,
) => {
    handle! fn_0 {0((v_1) => k_2) => {
        v_1 ++ <self>(k_2);
    }, pure(a_1) => "end"};
}

// .ts
const hash_563322b8: (arg_0: (handlers: any, done: (result: void) => void) => void) => string = fn_0 => {
  let handle_ret_0;
  handleSimpleShallow2("4e128638", fn_0, [(handlers, v_1, k_2) => {
    handle_ret_0 = v_1 + hash_563322b8(k_2);
  }], (handlers, a_1) => {
    handle_ret_0 = "end";
  });
  return handle_ret_0;
};

// .jd
hash_563322b8_0(
    () => {
        raise!(hash_4e128638_0("one"));
        raise!(hash_4e128638_0("two"));
        raise!(hash_4e128638_0("three"));
    },
) == "onetwothreeend"

// .ts
hash_563322b8((handlers: any, done: (arg_0: void) => void) => {
  raise(handlers, "4e128638", 0, "one", (handlers, value) => ((handlers, _ignored) => {
    raise(handlers, "4e128638", 0, "two", (handlers, value) => ((handlers, _ignored) => {
      raise(handlers, "4e128638", 0, "three", (handlers, value) => done(handlers, value));
    })(handlers, value));
  })(handlers, value));
}) == "onetwothreeend"

// .jd
hash_0b6f1408_0(hash_563322b8_0(() => {
    hash_596313f2_0("HI");
})) == "HIend"

// .ts
hash_0b6f1408(hash_563322b8((handlers: any, done: (arg_0: void) => void) => {
  hash_596313f2("HI", handlers, done);
})) == "HIend"

// .jd
hash_0b6f1408_0(hash_563322b8_0(() => {
    hash_596313f2_0("HI");
    hash_596313f2_0("HO");
})) == "HIHOend"

// .ts
hash_0b6f1408(hash_563322b8((handlers: any, done: (arg_0: void) => void) => {
  hash_596313f2("HI", handlers, (handlers, _ignored) => {
    hash_596313f2("HO", handlers, done);
  });
})) == "HIHOend"

// .jd
hash_0b6f1408_0(
    hash_563322b8_0(() => {
        raise!(hash_4e128638_0("HI"));
        raise!(hash_4e128638_0("HO"));
    }),
) == "HIHOend"

// .ts
hash_0b6f1408(hash_563322b8((handlers: any, done: (arg_0: void) => void) => {
  raise(handlers, "4e128638", 0, "HI", (handlers, value) => ((handlers, _ignored) => {
    raise(handlers, "4e128638", 0, "HO", (handlers, value) => done(handlers, value));
  })(handlers, value));
})) == "HIHOend"

// printFullName.jd
const hash_7359c246_0: () ={hash_4e128638_0, hash_762a49fa_0}> void = () => {
    hash_596313f2_0("What is your forename?");
    const foreName_0 = raise!(hash_762a49fa_0());
    hash_596313f2_0("What is your surname?");
    const surName_1 = raise!(hash_762a49fa_0());
    hash_596313f2_0(foreName_0 ++ " " ++ surName_1);
}

// .ts
const hash_7359c246: (handlers: any, done: (result: void) => void) => void = (handlers: any, done: (arg_0: void) => void) => {
  hash_596313f2("What is your forename?", handlers, (handlers, _ignored) => {
    raise(handlers, "762a49fa", 0, null, (handlers, value) => ((handlers, foreName_0) => ((handlers, _ignored) => {
      hash_596313f2("What is your surname?", handlers, (handlers, _ignored) => {
        raise(handlers, "762a49fa", 0, null, (handlers, value) => ((handlers, surName_1) => ((handlers, _ignored) => {
          hash_596313f2(foreName_0 + " " + surName_1, handlers, done);
        })(handlers))(handlers, value));
      });
    })(handlers))(handlers, value));
  });
};

// .jd
hash_0b6f1408_0(
    hash_563322b8_0(
        () => hash_5d8d14a2_0(
            "hi",
            () => hash_596313f2_0(raise!(hash_762a49fa_0()) ++ " and " ++ raise!(hash_762a49fa_0())),
        ),
    ),
) == "hi and hiend"

// .ts
hash_0b6f1408(hash_563322b8((handlers: any, done: (arg_0: void) => void) => {
  hash_5d8d14a2("hi", (handlers: any, done: (arg_0: void) => void) => {
    raise(handlers, "762a49fa", 0, null, (handlers, value) => ((handlers, arg_0_6) => {
      ((handlers, arg_0_5) => {
        raise(handlers, "762a49fa", 0, null, (handlers, value) => ((handlers, arg_1_5) => {
          ((handlers, arg_0_4) => {
            hash_596313f2(arg_0_4, handlers, done);
          })(handlers, arg_0_5 + arg_1_5);
        })(handlers, value));
      })(handlers, arg_0_6 + " and ");
    })(handlers, value));
  }, handlers, done);
})) == "hi and hiend"

// .jd
hash_0b6f1408_0(hash_563322b8_0(() => {
    hash_596313f2_0("A");
    hash_596313f2_0("B");
})) == "ABend"

// .ts
hash_0b6f1408(hash_563322b8((handlers: any, done: (arg_0: void) => void) => {
  hash_596313f2("A", handlers, (handlers, _ignored) => {
    hash_596313f2("B", handlers, done);
  });
})) == "ABend"

// .jd
hash_0b6f1408_0(
    hash_563322b8_0(
        () => hash_5d8d14a2_0(
            "hi",
            () => {
                hash_596313f2_0(hash_79d200c4_0());
                hash_596313f2_0("YA");
                hash_596313f2_0(hash_79d200c4_0());
                hash_596313f2_0("B");
            },
        ),
    ),
) == "hiYAhiBend"

// .ts
hash_0b6f1408(hash_563322b8((handlers: any, done: (arg_0: void) => void) => {
  hash_5d8d14a2("hi", (handlers: any, done: (arg_0: void) => void) => {
    hash_79d200c4(handlers, (handlers, arg_0_14) => {
      hash_596313f2(arg_0_14, handlers, (handlers, _ignored) => {
        hash_596313f2("YA", handlers, (handlers, _ignored) => {
          hash_79d200c4(handlers, (handlers, arg_0_11) => {
            hash_596313f2(arg_0_11, handlers, (handlers, _ignored) => {
              hash_596313f2("B", handlers, done);
            });
          });
        });
      });
    });
  }, handlers, done);
})) == "hiYAhiBend"

// .jd
hash_0b6f1408_0(
    hash_563322b8_0(
        () => hash_5d8d14a2_0(
            "hi",
            () => {
                raise!(hash_4e128638_0("A"));
                const x_0 = raise!(hash_762a49fa_0());
                raise!(hash_4e128638_0("B"));
            },
        ),
    ),
) == "ABend"

// .ts
hash_0b6f1408(hash_563322b8((handlers: any, done: (arg_0: void) => void) => {
  hash_5d8d14a2("hi", (handlers: any, done: (arg_0: void) => void) => {
    raise(handlers, "4e128638", 0, "A", (handlers, value) => ((handlers, _ignored) => {
      raise(handlers, "762a49fa", 0, null, (handlers, value) => ((handlers, x_0) => ((handlers, _ignored) => {
        raise(handlers, "4e128638", 0, "B", (handlers, value) => done(handlers, value));
      })(handlers))(handlers, value));
    })(handlers, value));
  }, handlers, done);
})) == "ABend"

// .jd
hash_0b6f1408_0(
    hash_563322b8_0(
        () => hash_5d8d14a2_0(
            "hi",
            () => {
                const x_0 = raise!(hash_762a49fa_0());
                raise!(hash_4e128638_0("B"));
            },
        ),
    ),
) == "Bend"

// .ts
hash_0b6f1408(hash_563322b8((handlers: any, done: (arg_0: void) => void) => {
  hash_5d8d14a2("hi", (handlers: any, done: (arg_0: void) => void) => {
    raise(handlers, "762a49fa", 0, null, (handlers, value) => ((handlers, x_0) => ((handlers, _ignored) => {
      raise(handlers, "4e128638", 0, "B", (handlers, value) => done(handlers, value));
    })(handlers))(handlers, value));
  }, handlers, done);
})) == "Bend"

// .jd
hash_0b6f1408_0(hash_563322b8_0(() => {
    hash_5d8d14a2_0("Me", hash_7359c246_0);
})) == "What is your forename?What is your surname?Me Meend"

// .ts
hash_0b6f1408(hash_563322b8((handlers: any, done: (arg_0: void) => void) => {
  hash_5d8d14a2("Me", hash_7359c246, handlers, done);
})) == "What is your forename?What is your surname?Me Meend"
`;

exports[`Example files generics.jd 1`] = `
// identity.jd
const hash_ef13ad00_0: (T_0) ={}> T_0 = (x_0: T_0) => x_0

// .ts
const hash_ef13ad00: <T_0>(arg_0: T_0) => T_0 = x_0 => x_0;

// get2.jd
const hash_2a7f74d2_0: int = hash_ef13ad00_0(2)

// .ts
const hash_2a7f74d2: number = hash_ef13ad00(2);

// callme.jd
const hash_8af07dc8_0: ((T_0) ={}> T_0, T_0) ={}> T_0 = (x_0: (T_0) ={}> T_0, arg_1: T_0) => x_0(
    arg_1,
)

// .ts
const hash_8af07dc8: <T_0>(arg_0: (arg_0: T_0) => T_0, arg_1: T_0) => T_0 = (x_0, arg_1) => x_0(arg_1);

// addTwo.jd
const hash_7987336f_0: (int) ={}> int = (one_0: int) => one_0 + 2

// .ts
const hash_7987336f: (arg_0: number) => number = one_0 => one_0 + 2;

// callIt.jd
const hash_6be6ce9a_0: int = hash_8af07dc8_0(hash_7987336f_0, 5)

// .ts
const hash_6be6ce9a: number = hash_8af07dc8(hash_7987336f, 5);

// scopeIt.jd
const hash_76f59cfc_0: (int) ={}> (int) ={}> int = (x_0: int) => (y_1: int) => x_0 + y_1

// .ts
const hash_76f59cfc: (arg_0: number) => (arg_0: number) => number = x_0 => y_1 => x_0 + y_1;

// ten.jd
const hash_7fc42012_0: int = hash_76f59cfc_0(4)(6)

// .ts
const hash_7fc42012: number = hash_76f59cfc(4)(6);
`;

exports[`Example files ifs.jd 1`] = `
// x.jd
const hash_cdfc6438_0: int = if 10 > 2 {
    5;
} else {
    4;
}

// .ts
const hash_cdfc6438: number = (() => {
  if (10 > 2) {
    return 5;
  } else {
    return 4;
  }
})();

// y.jd
const hash_220633d2_0: (int) ={}> int = (h_0: int) => if h_0 < 10 {
    h_0 / 2;
} else {
    h_0 * 3;
}

// .ts
const hash_220633d2: (arg_0: number) => number = h_0 => {
  if (h_0 < 10) {
    return h_0 / 2;
  } else {
    return h_0 * 3;
  }
};
`;

exports[`Example files inference.jd 1`] = `
// callOne.jd
const hash_44a7a53d_0: ((int, string) ={}> T_0) ={}> T_0 = (f_0: (int, string) ={}> T_0) => f_0(
    2,
    "hi",
)

// .ts
const hash_44a7a53d: <T_0>(arg_0: (arg_0: number, arg_1: string) => T_0) => T_0 = f_0 => f_0(2, "hi");

// inferrRecur.jd
const hash_8fed384a_0: (int) ={}> int = (count_0: int) => <self>(count_0 + 4)

// .ts
const hash_8fed384a: (arg_0: number) => number = count_0 => hash_8fed384a(count_0 + 4);
`;

exports[`Example files lets.jd 1`] = `
// x.jd
const hash_2ffe85a4_0: int = {
    const y_0 = 10;
    y_0;
}

// .ts
const hash_2ffe85a4: number = (() => {
  const y_0 = 10;
  return y_0;
})();

// z.jd
const hash_2813f64c_0: (int) ={}> int = (n_0: int) => {
    const m_1 = n_0 + 2;
    m_1 - 1;
}

// .ts
const hash_2813f64c: (arg_0: number) => number = n_0 => {
  const m_1 = n_0 + 2;
  return m_1 - 1;
};
`;

exports[`Type errors Expression 0 3 == "4" 1`] = `"first arg to == wrong type"`;

exports[`Type errors Expression 2 hello(5) 1`] = `
"Wrong type for arg 0: 
Found: int
but expected string : {\\"start\\":{\\"offset\\":110,\\"line\\":7,\\"column\\":1},\\"end\\":{\\"offset\\":118,\\"line\\":7,\\"column\\":9}}"
`;

exports[`Type errors Expression 4 count(hello("m")) 1`] = `
"Wrong type for arg 0: 
Found: string
but expected int : {\\"start\\":{\\"offset\\":152,\\"line\\":11,\\"column\\":1},\\"end\\":{\\"offset\\":169,\\"line\\":11,\\"column\\":18}}"
`;

exports[`Type errors Expression 6 identity<int>("hello") 1`] = `
"Wrong type for arg 0: 
Found: string
but expected int : {\\"start\\":{\\"offset\\":243,\\"line\\":18,\\"column\\":1},\\"end\\":{\\"offset\\":265,\\"line\\":18,\\"column\\":23}}"
`;

exports[`Type errors Expression 7 count(identity<string>("hello")) 1`] = `
"Wrong type for arg 0: 
Found: string
but expected int : {\\"start\\":{\\"offset\\":286,\\"line\\":21,\\"column\\":1},\\"end\\":{\\"offset\\":318,\\"line\\":21,\\"column\\":33}}"
`;

exports[`Type errors Expression 8 <T,>(x: T): int => x 1`] = `"Explicit type variable T#0 can't unify with int"`;

exports[`Type errors Expression 9 <T,>(x: int): T => x 1`] = `"Unable to unify int with type variable T#0"`;

exports[`Type errors Expression 11 () ={}> raise!(GetString.get()) 1`] = `"Function is declared as pure, but is not. 1 effects found."`;

exports[`Type errors Expression 12 <T>(ok: T): string => hello(ok) 1`] = `"Explicit type variable T#0 can't unify with string"`;

exports[`Type errors Expression 13 (ok: string) => ok + 2 1`] = `"first arg to + wrong type"`;

exports[`Type errors Expression 14 () => {
    const x = "hello";
    x + 2
} 1`] = `"first arg to + wrong type"`;

exports[`Type errors Expression 15 (fn: (string) => string) => {
    const x = fn(4);
    x ++ "hi"
} 1`] = `
"Wrong type for arg 0: 
Found: int
but expected string : {\\"start\\":{\\"offset\\":649,\\"line\\":46,\\"column\\":15},\\"end\\":{\\"offset\\":654,\\"line\\":46,\\"column\\":20}}"
`;

exports[`Type errors Expression 16 (fn: () => string) => {
    const x: int = fn();
    x + 2
} 1`] = `"Value of const doesn't match type annotation. string, expected int"`;

exports[`Type errors Expression 17 (fn: () => string) => {
    const x = fn();
    x + 2
} 1`] = `"first arg to + wrong type"`;
